### PR TITLE
Fork and reformat ILSVRC repo for RL use case, implement cartpole-v0

### PR DESCRIPTION
**Pull request by ghost**
We created a repo with shared abstractions, configurations, and conventions for training models on ILSVRC data (https://github.com/dbulhosa/ILSVRC). We are using the current state of this repo as a starting point for a new repo for training reinforcement learning agents. We will lean heavily on generators to simulate different agents and environments. For more information see the new README.

For our initial commit we:
- Deleted directory storing constants computed from Imagenet, which are not applicable to this new project. (`/shared/constants`)
- Deleted the testing runner which was very specific to testing image classifiers on multiple croppings of the same image. (`/shared/runners/test.py`)
- Deleted scripts and folders for downloading and pre-processing Imagenet images. (`/imagenet`)
- Deleted extraneous models and only kept one model directory and files as a starting point. (`/models/alexnet` and `/models/resnet`)

We will be modifying and deleting other files as we go to fit the purpose of this project.

### CODE REVIEW

**Comment by ghost** [comment]
My main concern at this point in time is whether the core concept of using a generator to interface between the environment and the model will work. In particular:

- Can we really multi-thread within a generator and if so how?
- Can we really pass the model to the generator constructor if the generator itself gets passed to the model's fit method?

Starting with a very simple model and environment to test the core premise of this approach probably makes sense. If it doesn't work we need to rethink how we will approach this. We will likely have to rely on other more granular methods in the Keras model class to make this work. This would be closer to what was done in `keras-rl`.

**Comment by ghost** [comment]
We think the best starting point is looking at the three classic discrete action RL environments offered by `gym` and training deep multi-layer perceptrons to solve them. These are:

- `MountainCar-v0`
- `CartPole-v1`
-  `Acrobot-v1`

We have actually solved the first ourselves in an exercise in a different repo: https://github.com/dbulhosa/suttonandbarto/blob/master/%5BUnfinished%5D%20S%26B%20RL%20Examples%209.1%2C%2010.1%2C%2012.2%2C%20%26%2013.1.ipynb

For these type of problems tile coding and linear modeling is a very intuitive and effective approach. However since we want to shift complexity to the model instead of custom feature engineering we will rely on DNNs instead.

**Comment by ghost** [comment]
Would be cool to have a callback that evaluates the agent at the end of every epoch. Should try implementing that after we have a simple model for classical control that runs with the custom sequence we built.

**Comment by ghost** [comment]
Environment generator now working. Need to test whether the actual training runner works next.

**Comment by ghost** [comment]
Last set of updates have gotten the code to where it runs!!! Yay!!! Next we need to debug for non-code errors. We need to actually train this agent for several epochs and see if it actually learns anything. Adding callbacks that measure the performance of the agent after each epoch and record it in the Tensorboard would help a lot both with this agent (classic control) and more complex (Atari) agents.

**Comment by ghost** [comment]
Fixed a couple of errors stopping the model from running multiple epochs. It runs multiple epochs now with the loss decreasing near monotonically most of the time! 

Callbacks measuring and logging agent performance are next. Also want to add a learning rate scheduler after to see how it affects performance.

**Comment by ghost** [comment]
Callbacks have been added. We also fixed a bug: We had forgotten to add logic that allows the agent to explore instead of taking the greedy action. We've allowed greediness to be scheduled. We can now successfully run the model and training, *and* we can log custom performance metrics to Tensorboard! 

Next we can focus on looking at whether the agent is learning anything, and fix silent errors. Running the model with metrics once we've noticed that the average reward for the agent per epoch is quite noisy. The loss also tends to go up and down aggressively, rather than mostly going down monotonically. We need to understand why this is happening. We should compare the performance of our agent with the benchmarks posted in the OpenAI Gym website.

**Comment by ghost** [comment]
Running into issue where: 

- Average reward evaluated after epoch goes up and down unpredictably. **[FIXED - Target model update with deep copy seems to have solved this, it now converges but to the wrong value]**
- Same is true of average initial estimated max value. **[FIXED - Same as above]**
- The difference between the average evaluate reward and average initial estimate max value is huge (an order of magnitude). **This would suggest there is something wrong with how the value function is being computed.**
- Before and after fixes the average total rewards per episode during evaluation have converged to ~9 and have not surpassed 20. For reference, we are expected to reach 195 ([see Leaderboard](https://github.com/openai/gym/wiki/Leaderboard)).

We tried:
- Using a more complex problem for the model, to make sure it's powerful enough to represent the problem.
- Initializing kernel weights randomly, to avoid symmetric initialization issues with the network weights messing up the gradient.
- Only starting teaching the model after N iterations using the random policy (this is from the paper), to stabilize training by making replay buffer distribution smoother.
- Increasing the replay buffer size significantly, for the same reason as the previous point.
- Reducing the learning rate (something that works when we see loss oscillation in DLCV), to see if this would eliminate the learning loss plateau.
- Change the loss of the final layer from `relu` to `linear` since this is the best practice for regression problems, which this model is solving.
- Increase the number of episodes simulated and states sampled in the evaluation callback, to reduce variance in performance estimates.

So far none of these have worked.... Some other ideas:

- Increase epsilon used during training. **[TRIED - NO EFFECT]**
- Build a problem that can be solved with tabular methods (using a custom environment) and see if we can even solve that:
    - One possibility is an environment that can be represented in a way that each model weight corresponds to a single state.
    - A simpler option is an environment with a single action and a couple of states to see if we can even estimate the value right, when actions are not really a factor.
- Check code for errors in (**note many of these can be checked by simply looking at an instance of the `EnvironmentGenerator`**):
    - How the loss is being computed. **[CHECKED - `model.clone_model` in Keras seems to not have been creating an actual full new copy of the current model for the target model. We started using `copy.deepcopy` instead and it worked!!! We figured the target model may be a problem because of the aforementioned observation that the value function did not match the average rewards. It does much better now.]**
    - Whether different parts of the generator are running as expected (by printing out)
    - That the replay buffer and observation buffers in general have the correct size, samples, etc. **[CHECKED - Found correct length]**
    - That our sampling method is actually working correctly. **[CHECKED - Indices sampled looked correct]**
    - That we didn't misalign states, actions,rewards, etc. in the buffers. **[CHECKED - Update buffer sizes to all be same so they aligned. All seemed to align after the change.]**
    - That the policy is selecting actions correctly **[TRIED - Improved agent performance but not all the way]***
- Increasing the frequency with which we update the target model. **[TRIED - NO EFFECT]**

Other things tried:

- Shorten epoch, evaluate agent before and after first epoch. **Found that random policy performs at about 9 reward average, so we knew agent wasn't learning.**

**Given what we're seeing and how pervasive the error is it seems like it's probably something very basic with the buffer, state sampling, or value function and its updates. Start by looking at these and understanding discrepancy between value function prediction and average rewards in evaluation.**

**Comment by ghost** [comment]
**FIXME**: At some point need to reformat learning schedule method calls in environment class so that we subtract the initial iterations from the actual iteration number.

**Comment by ghost** [comment]
Was able to get value predictions to match average rewards at evaluation. Also fixed an error with greedy action selection, should be correct now...

Now we're seeing two things:
+ The average reward and value prediction both converge to basically the performance of the random policy (~9). **[CHECKED - it's not learning the random policy, it's learning the constant policy which performs worse than the random one]**
+ The loss output by Keras is much higher (and does not seem to be the same) as the loss we get by computing the TD difference directly. **[CHECKED - They seem about equal now we think they're probably ok]**
+ The loss often increases quite quickly at some point mid epoch. This was not as much the case when the replay buffer was smaller and when the target model update was less frequent. **[CHECKED - Loss increase is mainly because Q function fitted does NOT fit terminal reward values, and increases uniformly on all action-state pairs]***

The latter two are quite problematic, we need to dig into it and figure out what's going on. They may be related. Also it may just be me but it seems that in certain minibatches the training freezes for a bit? What's the deal with that? **[CHECKED - Minibatch slowdown was because of overly frequent target model updates]**

Another interesting idea is to add the "average episode length" as a metric we write out to Tensorboard. **[DONE - done in the commits below, pretty helpful]**

Tried some things:
+ Increasing the number of iterations before updating target model, increasing replay buffer size: **[CHECKED - slows down loss divergence, but does not eliminate it. Does not increase performance. Did solve minibatch freeze -- it was probably due to frequent copies of model]**
+ Increase frequency of action selection: **[CHECKED - didn't fix anything]**
+ Check greedy actions are being selected at the right frequency: **[CHECKED - they are, so if there's a problem with action selection it's in the Q selection]**
+ Print out manually calculated losses, and loss per minibatch point: **[CHECKED - loss increase seems to come from model failing to predict reduced reward towards end of episodes. Average value function increases but it does so both on terminal states and non-terminal states uniformly. Did we implement our loss, value function, or something else algorithmic wrong?]**
+ Check indexing of `Q[a]` is model: **[CHECKED - is correct. Problem is not in extracting right `Q[a]`]**
+ Check state alignment: **[CHECKED - is correct. Seemed off at first but it's because index sampling breaks state sequence, so sequence arrays are not offsets of each other]**

The commits below reflect these leanings and changes. The next comment lays out next steps.

**Comment by dbulhosa** [comment]
We're actually performing worse than the random policy. The random policy should be performing at about ~19.36 average reward with an average episode length of ~22.31.

**Comment by ghost** [comment]
We're actually performing worse than the random policy. The random policy should be performing at about ~19.36 average reward with an average episode length of ~22.31. It turns out that if the weight initialization is anything but constant the initial policy is not random. Furthermore, even when the initial policy is random the learned policy is worse than the random policy. The policy learned in fact is the constant policy. 

Current problems observed are these, which are just better diagnoses of previously observed issues:

+ The model seems to learn to prefer one action universally (it can be either, depends on target model update).
+ Because of this, the model learns a worse policy than random action choice. We first suspected the policy learned was the constant policy because of the worse performance.
+ The Q function fitted seems to be near uniform and does not fit at all the terminal values.
+ The difference between the Q values of the two different actions is almost exactly the same across all different states passed to it. The fact there is a constant value difference between the two actions is a very specific problem. **It's almost as if `Q[0]` and `Q[1]` are being updated by the same amount... Like the gradient for one action is being applied to both action values?**
+ These problems persist even if we use the current model as the target model (i.e. no delay). Thus we don't think it's a model copying issue.

These issues may be related... We suspect it's something with the Keras model or the loss but we are not sure what exactly. Should walk through model in detail, we think these last problems represent the final stretch though, so we're getting there!!!

**Comment by ghost** [comment]
We figured out how to print out weights and gradients for the model. We're running into an issue where the MSE we're calculating is totally wrong. The SE used to calculate the MSE is two-dimensional where it should be one-dimensional. We're waiting to draw conclusions about whether the problem is with the gradient or not after we fix this problem and see if the resulting gradients make sense or not.

**Comment by ghost** [comment]
We figured out the issue: the model output `Q_a` (which is the second model output) had the wrong shape. It likely had shape `(minibatch_size, )` instead of `(minibatch_size, 1)`. This was causing a broadcasting error that was in turn causing the loss `(y - Q_a)` to be calculated incorrectly! We fixed the shape both of the model output and of the generator output and this fixed the whole thing!!! All of the aforementioned issues have been solved by the change and we were able to train a model solving CartPole that successfully learns it!!!

We've cleaned the PR by eliminating unused files. We'll readd them in one of the next PRs where we add DQN. Next we will create a PR to optimize the parameters for this problem and use the same (or similar) models to solve other discrete classical control problems. After we are confident and comfortable with our DQN solutions for these simple problems we'll do a new PR to implement DQN for Atari.

**Comment by ghost** [comment]
We checked the checkpoints and videos, and our agent in fact did solve the problem! We found that with time the agents solution got more sophisticated, but that the agents performance was incredibly noisy. We think part of this was probably due to the simplicity of the model we used. More parameters should make it easier for the agent to evolve its policy without dramatically reducing its performance between learning steps. Next we move on to hyperparameter tuning and other discrete classic control problems.
